lib/sla.js

```diff
@@ -40,54 +41,108 @@ function designateMixingBridge(client, bridgeName) {
  * @param {Object} bridge - Bridge that channels are to be added to.
+ * @param {Object} extension - the object data related to this sharedExtension
  * @return {Q} Q promise object.
  */
-function addChannelsToBridge(inbound, dialed, bridge) {
+function addChannelsToBridge(client, channel, dialed, bridge, extension) {
+  // Resets listeners, so that redundant events aren't accepted.
+  bridge.removeAllListeners('ChannelEnteredBridge');
+  bridge.removeAllListeners('ChannelLeftBridge');
   var defer = Q.defer();
-  var numInBridge = 0;
   console.log('Adding channels to bridge %s', bridge.id);
   var addChannel = Q.denodeify(bridge.addChannel.bind(bridge));
-  addChannel({channel: [inbound.id, dialed.id]})
+  addChannel({channel: channel.id})
     .catch(function (err) {
       defer.reject(err);
     });
+  if (dialed) {
+    addChannel({channel: dialed.id})
+      .catch(function (err) {
+        defer.reject(err);
+      });
+  }
+  /**
+   * Function that displays a helpful message when a channel enters a bridge
+   * @param {Object} event - the event related to this bridge entering
+   * @param {Object} object - contains the bridge and channel in question
+   */
   function channelEntered(event, object) {
     console.log('Channel %s has entered the bridge', object.channel.id);
-    numInBridge += 1;
   }
   bridge.on('ChannelEnteredBridge', channelEntered);
+  /**
+   * Function that is utilized when a channel leaves the bridge.
+   *   If the channel is a non-station channel, then hangs up remaining
+   *   channels.  If the channel is a station, checks if the only non-station
+   *   channel is the only left, and then hangs up that non-station channel.
+   * @param {Object} event - the event related to the bridge leaving
+   * @param {Object} object - contains the bridge and channel in question
+   */
   function channelLeft(event, object) {
-    numInBridge -= 1;
-    console.log('Channel %s has left the bridge', object.channel.id);
-    if (numInBridge === 1) {
-      if (object.channel.id === inbound.id) {
-        console.log('Hanging up dialed channel %s', dialed.id);
-        var hangupDialed = Q.denodeify(dialed.hangup.bind(dialed));
-        hangupDialed()
-          .catch(function (err) {
-            err.name = 'HangupFailure';
-            defer.reject(err);
+    isStation(extension, object.channel).then(function (isStation) {
+      if (isStation) {
+        if (object.bridge.channels.length === 1) {
+          var hangupOutsider = Q.denodeify(client.channels.hangup.bind
+              (client.channels));
+          hangupOutsider({channelId: object.bridge.channels[0]})
+            .catch(function (err) {
+              err.name = 'HangupFailure';
+              defer.reject(err);
+            });
+        }
+      } else {
+        if (object.bridge.channels.length) {
+          console.log('Hanging up station channel/s');
+          object.bridge.channels.forEach( function(ID) {
```

Comment:
This will work as expected but you may want to look at Q.all() to run the hangups concurrently. It will then call then once all have been completed or catch if a single async operation failed.